### PR TITLE
Avoid full reapplication of `cache-and-network` and `network-only` fetch policies after successful `fetchMore`

### PR DESCRIPTION
**Pull request by benjamn**
This PR implements a variation on the idea I sketched in https://github.com/apollographql/apollo-client/issues/6916#issuecomment-901466918, ~without doing too much violence to our existing test suite~. I did end up making a number of adjustments to tests that this PR initially caused to fail, but I'm confident they are nonviolent (that is, they test the same things in a more robust/realistic way).

Previous attempts to solve #6916 stalled in part because the number of test failures felt indicative of potential backwards incompatibility. Now that I've worked through all the failing tests in this PR, I believe they are mostly just fragile/flaky/timing-sensitive tests, though of course it would be great to validate that hope with beta testing.

In short, `fetchMore` should (now) be able to deliver complete cache results to `ObservableQuery` subscribers without triggering additional network requests when using the `cache-and-network` or `network-only` fetch policies, resolving the following related (some long-standing) issues:
* #6916
* #7939
* #9363
* https://github.com/vuejs/apollo/issues/1026
* … and likely more

### CODE REVIEW

File: src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -746,19 +758,56 @@ once, rather than every time you call fetchMore.`);
...
+    if (
+      fetchPolicy === "cache-and-network" ||
+      fetchPolicy === "network-only"
+    ) {
+      return this.reobserve({
+        fetchPolicy: "cache-first",
+        // Use a temporary nextFetchPolicy function that replaces itself with
+        // the previous nextFetchPolicy value and returns the original
+        // fetchPolicy.
+        nextFetchPolicy(...args) {
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:793]: The key to avoiding test failures was to embrace/reuse the `ObservableQuery#reobserve` machinery (not the more basic and private `observe` or `reportResult` methods), temporarily substituting a `cache-first` fetch policy in place of `cache-and-network` or `network-only` policies.

A temporary `nextFetchPolicy` function restores the original `fetchPolicy` and `nextFetchPolicy` immediately after this reobservation, leaving the `this.options` the same as they were before. That's the idea, at least.

The point is twofold:
1. reuse the `ObservableQuery` reobservation pipeline, which handles some legacy functionality like `@client @export(as:...)` directives
2. allow _complete_ cache results to be delivered to `ObservableQuery` subscribers without kicking off (almost always unwanted) network requests

Alternatively, we could change the global default for `nextFetchPolicy` to `cache-first` (or `cache-first` specifically for `cache-and-network` and `network-only` initial policies). In that approach, the original `options.fetchPolicy` changes permanently to `cache-first`, whereas in this approach the original `fetchPolicy` stays the same, changing only momentarily to `cache-first` when cache results are delivered.

I'm not sure I have a strong preference between these two approaches at the moment, but happy to discuss further.

File: src/core/QueryInfo.ts [hunk 2]
```diff
@@ -234,8 +234,11 @@ export class QueryInfo {
         // full reobservation, since oq.reobserve might make a network
         // request, and we don't want to trigger network requests for
         // optimistic updates.
-        if (this.getDiff().fromOptimisticTransaction) {
+        const diff = this.getDiff();
+        if (diff.fromOptimisticTransaction) {
           oq["observe"]();
+        } else if (diff.complete) {
+          oq["reobserveCacheFirst"]();
```
Comment (benjamn) [inline src/core/QueryInfo.ts:241]: This change (delivering complete cache results with `reobserveCacheFirst`) also seems like a good idea, though it's not strictly necessary solve the `fetchMore` problem, because `fetchMore` calls `reobserveCacheFirst` independently.

File: src/core/ObservableQuery.ts [hunk 3]
```diff
@@ -430,14 +432,29 @@ once, rather than every time you call fetchMore.`);
           query: combinedOptions.query,
           variables: combinedOptions.variables,
           data,
+          // TODO Figure out why this breaks tests, since the result should
+          // ultimately be broadcast by this.reobserveCacheFirst in the finally
+          // block below. Alternatively, we rely on the cache broadcast for this
+          // cache.writeQuery, and avoid using reobserveCacheFirst below.
+          broadcast: false,
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:439]: Converting back to a draft PR until I can investigate the interaction between this `cache.writeQuery` call/broadcast and the `this.reobserveCacheFirst()` call below. I thought disabling this cache broadcast should be safe now, but it caused several (12) test failures, so I think I need to keep investigating.

File: src/core/QueryInfo.ts (hunk 2 above)
Comment (benjamn) [reply src/core/QueryInfo.ts:241]: Update: this call to `reobserveCacheFirst` is now the primary mechanism by which `fetchMore` delivers results, with the call in the `finally` block serving as a backstop. So this change is pretty important now, contradicting what I said in my previous comment.

File: src/core/ObservableQuery.ts (hunk 3 above)
Comment (benjamn) [reply src/core/ObservableQuery.ts]: I investigated this further, and I now believe the cache broadcast that comes from `fetchMore` writing to the cache should be the signal that (usually) reports the new result (so it does not make sense to pass `broadcast: false` here), and `reobserveCacheFirst` should be called in the `finally` block in `fetchMore` only in cases where the cache broadcast did not happen.

File: src/core/ObservableQuery.ts [hunk 4]
```diff
@@ -381,64 +388,73 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`);
...
-Please convert updateQuery functions to field policies with appropriate
-read and merge functions, or use/adapt a helper function (such as
-concatPagination, offsetLimitPagination, or relayStylePagination) from
-@apollo/client/utilities.
-
-The field policy system handles pagination more effectively than a
-hand-written updateQuery function, and you only need to define the policy
-once, rather than every time you call fetchMore.`);
-          warnedAboutUpdateQuery = true;
-        }
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:419]: Removing this warning is something a number of developers have requested (for example, [this twitter discussion](https://twitter.com/zachcodes/status/1484919400893501440)). At this point, I believe the warning has done all the good it can do (nudging people to stop using `updateQuery` and use field policies instead), so removing the warning will improve the lives of developers who still have a good reason to use `updateQuery`. Also, `updateQuery` is now implemented in terms of `cache.updateQuery` (#8382), which is an API we intend to continue supporting.

**Comment by benjamn** [comment]
As a demonstration of the safety of the `src/__tests__/fetchMore.ts` test changes, I rebased commit bb83bca87d5f91565fda000a5bfdd397a239a566 all the way to the beginning of the branch, then force-pushed it by itself, so it's easy to see the modified tests passing in CI before reviewing the rest of the commits.

File: src/core/ObservableQuery.ts [hunk 5]
```diff
@@ -786,19 +811,56 @@ once, rather than every time you call fetchMore.`);
...
   }
 
+  // Reobserve with fetchPolicy effectively set to "cache-first", triggering
+  // delivery of any new data from the cache, possibly falling back to the
+  // network if any cache data are missing. This allows _complete_ cache results
+  // to be delivered without also kicking off unnecessary network requests when
+  // this.options.fetchPolicy is "cache-and-network" or "network-only". When
+  // this.options.fetchPolicy is any other policy ("cache-first", "cache-only",
+  // "standby", or "no-cache"), we call this.reobserve() as usual.
+  public reobserveCacheFirst() {
```
Comment (benjamn) [inline src/core/ObservableQuery.ts:834]: We could keep this method `private` for now, if we don't want to be stuck with it in the public `ObservableQuery` API. Thoughts?

Comment (benjamn) [inline src/core/ObservableQuery.ts:834]: We could keep this method `private` for now, if we don't want to be stuck with it in the public `ObservableQuery` API. Thoughts?

Comment (hwillson) [reply src/core/ObservableQuery.ts:834]: `private` sounds great!

File: src/core/ObservableQuery.ts (hunk 1 above)
Comment (hwillson) [reply src/core/ObservableQuery.ts:846]: We'll aim to change the `nextFetchPolicy` global default in 4.0.

File: src/core/ObservableQuery.ts (hunk 5 above)
Comment (benjamn) [reply src/core/ObservableQuery.ts]: https://github.com/apollographql/apollo-client/pull/9504/commits/5036dc531fc7a60dd091b82b5c8e44cef43fb488

**Comment by benjamn** [comment]
These changes have been published to npm as `@apollo/client@3.6.0-beta.10`.
